ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java

```diff
@@ -0,0 +1,99 @@
+package com.applovin.mediation;
+
+import android.content.Context;
+import android.graphics.drawable.Drawable;
+import android.net.Uri;
+import android.os.Bundle;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ImageView;
+
+import com.applovin.nativeAds.AppLovinNativeAd;
+import com.google.android.gms.ads.formats.NativeAd;
+import com.google.android.gms.ads.mediation.NativeAppInstallAdMapper;
+
+import java.util.ArrayList;
+
+public class AppLovinNativeAdMapper extends NativeAppInstallAdMapper {
+
+    private AppLovinNativeAd nativeAd;
+
+    public AppLovinNativeAdMapper(AppLovinNativeAd nativeAd, Context context) {
+        this.nativeAd = nativeAd;
+        setHeadline(nativeAd.getTitle());
+        setBody(nativeAd.getDescriptionText());
+        setCallToAction(nativeAd.getCtaText());
+
+        ImageView mediaView = new ImageView(context);
+        ViewGroup.LayoutParams layoutParams =
+                new ViewGroup.LayoutParams(
+                        ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
+        mediaView.setLayoutParams(layoutParams);
+
+        ArrayList<NativeAd.Image> images = new ArrayList<>(1);
+        AppLovinNativeAdImage image;
+        AppLovinNativeAdImage icon;
+        Drawable imageDrawable =
+                Drawable.createFromPath(Uri.parse(nativeAd.getImageUrl()).getPath());
+        Drawable iconDrawable = Drawable.createFromPath(Uri.parse(nativeAd.getIconUrl()).getPath());
+        image = new AppLovinNativeAdImage(Uri.parse(nativeAd.getImageUrl()), imageDrawable);
+        icon = new AppLovinNativeAdImage(Uri.parse(nativeAd.getIconUrl()), iconDrawable);
+        images.add(image);
+        setImages(images);
+        setIcon(icon);
+
+        mediaView.setImageDrawable(imageDrawable);
+        setMediaView(mediaView);
+        setStarRating(nativeAd.getStarRating());
+
+        Bundle extraAssets = new Bundle();
+        extraAssets.putLong(AppLovinNativeAdapter.KEY_EXTRA_AD_ID, nativeAd.getAdId());
+        extraAssets.putString(
+                AppLovinNativeAdapter.KEY_EXTRA_CAPTION_TEXT, nativeAd.getCaptionText());
```

Comment:
`Bundle` employs a builder pattern so you can technically do:

```
Bundle extraAssets = new Bundle().putLong(AppLovinNativeAdapter.KEY_EXTRA_AD_ID, nativeAd.getAdId())
     .putString(...)
    .putThat(...);
```

Reply:
I don't think that is true. [`putLong(key, value)`](https://developer.android.com/reference/android/os/BaseBundle.html#putLong(java.lang.String,%20long)) (or any other `Bundle.put...(key, value)`) is not returning a bundle, so we won't be able to chain multiple puts together.

Reply:
Ah I must've been thinking about something else..

Reply:
Intents!